Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,290 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- This file is included by "Data.ByteString.ReadInt",
+-- "Data.ByteString.ReadInt64", and "Data.ByteString.Lazy.ReadInt64", after
+-- defining "BYTESTRING_STRICT" or "API_BITS_64" or both.  The modules are
+-- largely identical, except for the choice of ByteString type, and the
+-- the use of native words, or explicit 64-bits words for the accumulator.
+
+#ifndef API_BITS_64
+
+# ifdef BYTESTRING_STRICT
+module Data.ByteString.ReadInt
+# else
+module Data.ByteString.Lazy.ReadInt
+# endif
+    ( readInt
+    , readInt8
+    , readInt16
+    , readInt32
+    , readWord
+    , readWord8
+    , readWord16
+    , readWord32
+    ) where
+
+#else
+
+# ifdef BYTESTRING_STRICT
+module Data.ByteString.ReadInt64
+# else
+module Data.ByteString.Lazy.ReadInt64
+# endif
+    ( readInt64, readWord64 ) where
+
+#endif
+
+import qualified Data.ByteString.Internal as BI
+#ifdef BYTESTRING_STRICT
+import Data.ByteString as B
+#else
+import Data.ByteString.Lazy as B
+import Data.ByteString.Lazy.Internal
+#endif
+import Data.Int
+import Data.Word
+import Foreign.ForeignPtr (ForeignPtr)
+import Foreign.Ptr (minusPtr, plusPtr)
+import Foreign.Storable (Storable(..))
+
+----- Public API
+
+#ifndef API_BITS_64
+type Accum = Word
```

Comment:
There is no architecture-dependence here. `API_BITS_64` is defined in ReadInt64.hs to make sure the accumulator used in `readInt64` and `readWord64` is wide enough on 32-bit architecture, while all others can just use `Word`. (See also the comments at the top of this file.)

Reply:
Use of a native accumulator type simplifies the code even on 64-bit systems, where otherwise with GHC 9.2, we still end up with various additional primops coverting `Word64#` to `Word#` and back, though perhaps ultimately these might generate no code...

But, if you feel strongly about it, yes we could just use Word64 everywhere, leading to less efficient 32-bit code.  I don't know how much use Haskell still has on 32-bit systems...

Reply:
I do not believe there are any performance-critical 32-bit applications in Haskell by now, much less by the time `bytestring-0.12` is released. If we can avoid `API_BITS_64` switch and `ReadInt64` modules by sacrificing 32-bit performance, by all means please go for it.

Reply:
No worries.  Dropped the API switch and squashed.